src/common/memory/MemoryUtils.cpp

```diff
@@ -118,18 +127,77 @@ StatusOr<bool> MemoryUtils::hitsHighWatermark() {
     }
   }
 
-  // MemoryStats depends on jemalloc
-#ifdef ENABLE_JEMALLOC
-#ifndef ENABLE_ASAN
-  // set MemoryStats limit (MemoryTracker track-able memory)
-  int64_t trackable = total - FLAGS_memory_tracker_untracked_reserved_memory_mb * MiB;
-  if (trackable > 0) {
-    MemoryStats::instance().setLimit(trackable * FLAGS_memory_tracker_limit_ratio);
+  handleMemoryTracker(total, available);
+
+  auto hits = (1 - available / total) > FLAGS_system_memory_high_watermark_ratio;
+  LOG_IF_EVERY_N(WARNING, hits, 100)
+      << "Memory usage has hit the high watermark of system, available: " << available
+      << " vs. total: " << total << " in bytes.";
+  return hits;
+}
+
+StatusOr<uint64_t> MemoryUtils::readSysContents(const std::string& path) {
+  std::ifstream ifs(path);
+  if (!ifs) {
+    return Status::Error("Could not open the file: %s", path.c_str());
+  }
+  uint64_t value = 0;
+  ifs >> value;
+  return value;
+}
+
+void MemoryUtils::handleMemoryTracker(int64_t total, int64_t available) {
+#ifdef ENABLE_MEMORY_TRACKER
+  double limitRatio = FLAGS_memory_tracker_limit_ratio;
+  double availableRatio = FLAGS_memory_tracker_available_ratio;
+
+  double untrackedMb = FLAGS_memory_tracker_untracked_reserved_memory_mb;
+
+  // update MemoryTracker limit memory by Flags
+  if (limitRatio > 0.0 && limitRatio <= 1.0) {
+    /**
+     * (<= 1.0): Normal limit is set to ratio of trackable memory;
+     *           limit = limitRatio * (total - untrackedMb)
+     */
+    if (kCurrentTotal_ != total || kCurrentLimitRatio_ != limitRatio) {
+      double trackable = total - (untrackedMb * MiB);
+      if (trackable > 0) {
+        MemoryStats::instance().setLimit(trackable * limitRatio);
+      } else {
+        LOG(ERROR) << "Total memory less than untracked " << untrackedMb << " Mib,"
+                   << " MemoryTracker limit is not set properly, "
+                   << " Current memory stats:" << MemoryStats::instance().toString();
+      }
+      LOG(INFO) << "MemoryTracker set static ratio: " << limitRatio;
+    }
+  } else if (limitRatio == 2.0) {
```

Comment:
best not to use equals to compare floating point data types

Reply:
> best not to use equals to compare floating point data types

I had considered it as a issue, but I tried in real run environment, it works as expected.